RealmSwift/Query.swift

```diff
@@ -0,0 +1,1075 @@
+        return lhs.append(expression: [.basicComparison(.equal), .rhs(rhs)])
+    }
+    /// :nodoc:
+    public static func != <V>(_ lhs: Query<V>, _ rhs: V) -> Query where V: _RealmSchemaDiscoverable {
+        return lhs.append(expression: [.basicComparison(.notEqual), .rhs(rhs)])
+    }
+
+    // MARK: Numerics
+
+    /// :nodoc:
+    public static func > <V>(_ lhs: Query<V>, _ rhs: V) -> Query where V: _QueryNumeric {
+        return lhs.append(expression: [.basicComparison(.greaterThan), .rhs(rhs)])
+    }
+    /// :nodoc:
+    public static func >= <V>(_ lhs: Query<V>, _ rhs: V) -> Query where V: _QueryNumeric {
+        return lhs.append(expression: [.basicComparison(.greaterThenOrEqual), .rhs(rhs)])
+    }
+    /// :nodoc:
+    public static func < <V>(_ lhs: Query<V>, _ rhs: V) -> Query where V: _QueryNumeric {
+        return lhs.append(expression: [.basicComparison(.lessThan), .rhs(rhs)])
+    }
+    /// :nodoc:
+    public static func <= <V>(_ lhs: Query<V>, _ rhs: V) -> Query where V: _QueryNumeric {
+        return lhs.append(expression: [.basicComparison(.lessThanOrEqual), .rhs(rhs)])
+    }
+
+    // MARK: Compound
+
+    /// :nodoc:
+    public static func && (_ lhs: Query, _ rhs: Query) -> Query {
+        // Wrap the left expression and right expression in parentheses
+        var copy = lhs
+        copy.context.expression.insert(.special(.openParentheses), at: 0)
+        return copy.append(expression: [.compound(.and)] + rhs.context.expression + [.special(.closeParentheses)],
+                           subqueryCount: rhs.context.count)
+    }
+    /// :nodoc:
+    public static func || (_ lhs: Query, _ rhs: Query) -> Query {
+        // Wrap the left expression and right expression in parentheses
+        var copy = lhs
+        copy.context.expression.insert(.special(.openParentheses), at: 0)
+        return copy.append(expression: [.compound(.or)] + rhs.context.expression + [.special(.closeParentheses)],
+                           subqueryCount: rhs.context.count)
+    }
+
+    // MARK: Subscript
+
+    /// :nodoc:
+    public subscript<V>(dynamicMember member: KeyPath<T, V>) -> Query<V> where T: ObjectBase {
+        let name = _name(for: member)
+        return append(expression: [.keyPath(name: name)])
+    }
+    /// :nodoc:
+    public subscript<V: RealmCollectionBase>(dynamicMember member: KeyPath<T, V>) -> Query<V> where T: ObjectBase {
+        let name = _name(for: member)
+        return append(expression: [.keyPath(name: name, isCollection: true)])
+    }
+
+    // MARK: Query Construction
+
+    /// Creates an NSPredicate compatibe string.
+    /// - Parameter isSubquery: States if expression need to be arraged in a special way to cater to subqueries.
+    /// - Returns: A tuple containing the predicate string and an array of arguments.
+    public func _constructPredicate(_ isSubquery: Bool = false) -> QueryPredicate {
```

Comment:
Creating a NSPredicate by serializing our representation to a string and then asking NSPredicate to parse that is pretty roundabout. This all seems like it'd be a lot simpler if it just constructed a NSPredicate.

Reply:
I agree, the reason for building the string out like this was because it offered more granular control. For example we need that granular control when construction a collection aggregate predicate– we need to transform `$0.arrayCol.intCol.avg` to `arrayCol.@avg.intCol` so we need to insert the `@avg` operator at a certain position of the string. I'll look into seeing if we can build the NSPredicate along the way instead of passing a large string to the NSPredicate constructor at the end.